### PR TITLE
Add demo showcasing QML optimization with Catalyst and Optax

### PR DESCRIPTION
**Pull request by josh146**

### CODE REVIEW

**Comment by josh146** [comment]
[sc-53769]

**Comment by josh146** [comment]
[sc-57415]

File: demos/qml_optimization.ipynb [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+      },
+      "outputs": [],
+      "source": [
+        "@qml.qnode(dev)\n",
+        "def circuit(data, weights):\n",
+        "    \"\"\"Quantum circuit ansatz\"\"\"\n",
+        "\n",
+        "    @catalyst.for_loop(0, n_wires, 1)\n",
+        "    def data_embedding(i):\n",
+        "        qml.RY(data[i], wires=i)\n",
```
Comment (dime10) [inline demos/qml_optimization.ipynb:68]: Is worth presenting some built-in PennyLane embeddings, or did you chose this specifically for the control flow? :)

File: demos/qml_optimization.ipynb [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+        "        qml.RX(weights[i, 2], wires=i)\n",
+        "        qml.CNOT(wires=[i, (i + 1) % n_wires])\n",
+        "\n",
+        "    ansatz()\n",
+        "\n",
+        "    # we use a sum of local Z's as an observable since a\n",
+        "    # local Z would only be affected by params on that qubit.\n",
+        "    return qml.expval(qml.sum(*[qml.PauliZ(i) for i in range(n_wires)]))\n",
+        "\n",
+        "circuit = qml.qjit(catalyst.vmap(circuit, in_axes=(1, None)), autograph=False)"
```
Comment (dime10) [inline demos/qml_optimization.ipynb:85]: Nice, putting all the features together 😄  I guess the only thing missing is autograph, is that because of specific issues right now, or because non-autograph is generally more robust?

File: demos/qml_optimization.ipynb [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+      "cell_type": "code",
+      "execution_count": 7,
+      "metadata": {
+        "id": "sOWgLW5esIsO"
+      },
+      "outputs": [],
+      "source": [
+        "weights = jnp.ones([n_wires, 3])\n",
+        "bias = jnp.array(0.)\n",
+        "params = {\"weights\": weights, \"bias\": bias}"
```
Comment (dime10) [inline demos/qml_optimization.ipynb:138]: Truly putting everything together 😆

File: demos/qml_optimization.ipynb (hunk 2 above)
Comment (dime10) [inline demos/qml_optimization.ipynb:85]: Is the batching on axis 1 intentional? (I think it will create some partial transposition compared to the "intuitive" result of vmapping)

File: demos/qml_optimization.ipynb [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "99kM0UDs83QM"
+      },
+      "source": [
+        "## JIT-compiling the optimization\n",
+        "\n",
+        "In the above example, we just-in-time (JIT) compiled our cost function `loss_fn`. However, we can also JIT compile the entire optimization loop; this means that the for-loop around optimization is not happening in Python, but is compiled and executed native. This avoids (potentially costly) data transfer between Python and our JIT compiled cost function with each update step."
```
Comment (dime10) [inline demos/qml_optimization.ipynb:305]:
```suggestion
        "In the above example, we just-in-time (JIT) compiled our cost function `loss_fn`. However, we can also JIT compile the entire optimization loop; this means that the for-loop around optimization is not happening in Python, but is compiled and executed natively. This avoids (potentially costly) data transfer between Python and our JIT compiled cost function with each update step."
```

File: demos/qml_optimization.ipynb [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+      ],
+      "source": [
+        "print(final_params)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## Timing the optimization"
```
Comment (dime10) [inline demos/qml_optimization.ipynb:377]: Should the demo show raw pennylane as well?

**Review by dime10 — APPROVED** [review]
Nice demo 👌

File: demos/qml_optimization.ipynb (hunk 2 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:85]: Unfortunately this is because of the issue that AutoGraph is converting the following if statement into a `catalyst.cond`:

https://github.com/PennyLaneAI/catalyst/blob/afbe7035ccb84cf5cbff2f0b1985e99349fb2e64/frontend/catalyst/pennylane_extensions.py#L2265-L2266

which is causing vmap to break (it tries to dispatch to JAX when it shouldn't).

This is the only thing stopping me from using Autograph!

Comment (josh146) [reply demos/qml_optimization.ipynb:85]: Batching axis of 1 is intentional, because I'm matching the results of this demo here to make sure I get the same results:

https://pennylane.ai/qml/demos/tutorial_How_to_optimize_QML_model_using_JAX_and_Optax/

Here, vmap isn't used; instead, operator vectorization is used directly:

image

The interesting thing about operator vectorization is that it is roughly equivalent to a batch dimension at the _end_ of the data, rather than the beginning! Because it's the 'last' axes of the data array that is provided to the operator.

File: demos/qml_optimization.ipynb (hunk 5 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:377]: Raw PL as in jax + default qubit, or raw PL as in lightning.qubit?

Interestingly:

- JAX + default.qubit is significantly faster, at 0.0031725019999839787 sec per loop
- lightning.qubit is significantly slower, at ~1 minute per loop.

I left out the lightning comparison because it was taking too long to time lol

File: demos/qml_optimization.ipynb (hunk 3 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:138]: yep, I needed the pytree support for this!

File: demos/qml_optimization.ipynb (hunk 1 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:68]: I chose this to match https://pennylane.ai/qml/demos/tutorial_How_to_optimize_QML_model_using_JAX_and_Optax/, although I can't remember why we chose it there 😅

File: demos/qml_optimization.ipynb [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+            "{'bias': array(-0.75432067), 'weights': array([[-1.95077271e-01,  5.28546590e-02, -4.89252073e-01],\n",
+            "       [-1.99687794e-02, -5.32871564e-02,  9.22904864e-02],\n",
+            "       [-2.71755507e-03, -9.64672786e-05, -4.79570827e-03],\n",
+            "       [-6.35443870e-02,  3.61110014e-02, -2.05196876e-01],\n",
+            "       [-9.02635405e-02,  1.63759364e-01, -5.64262612e-01]])}\n"
+          ]
+        }
+      ],
+      "source": [
+        "print(qml.qjit(catalyst.grad(loss_fn, method=\"fd\"))(params, data, targets))"
```
Comment (rmoyard) [inline demos/qml_optimization.ipynb:200]:
```suggestion
        "print(qml.qjit(catalyst.grad(loss_fn))(params, data, targets))"
```
When https://github.com/PennyLaneAI/catalyst/pull/332 is merged :+1:

File: demos/qml_optimization.ipynb (hunk 1 above)
Comment (dime10) [reply demos/qml_optimization.ipynb:68]: Got it 😅

File: demos/qml_optimization.ipynb (hunk 2 above)
Comment (dime10) [reply demos/qml_optimization.ipynb:85]: Oh interesting, good to know! To readers the `in_axes=(1, None)` might be coming out of the blue though.

Comment (dime10) [reply demos/qml_optimization.ipynb:85]: This example seems to work fine with autograph and vmap, there must be another element 🤔 
```py
@vmap
@qml.qnode(qml.device("lightning.qubit", wires=1))
def f(x):
    qml.RY(x, wires=0)
    return qml.probs()

@qjit(autograph=True)
def g():
    x = jnp.array([0., 1., 2., 3.])
    return f(x)

g()
```

File: demos/qml_optimization.ipynb [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+      "metadata": {
+        "id": "w4QPSXdysqJx"
+      },
+      "outputs": [],
+      "source": [
+        "@qml.qnode(dev)\n",
+        "def circuit(data, weights):\n",
+        "    \"\"\"Quantum circuit ansatz\"\"\"\n",
+        "\n",
+        "    @catalyst.for_loop(0, n_wires, 1)\n",
```
Comment (dime10) [inline demos/qml_optimization.ipynb:66]: The demo uses a mix of catalyst and qml prefixes for catalyst functions, for instance `catalyst.for_loop` but `qml.qjit`. Should be consistent here?

File: demos/qml_optimization.ipynb [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+      "cell_type": "code",
+      "execution_count": 3,
+      "metadata": {
+        "id": "iKtDMXjENR7X"
+      },
+      "outputs": [],
+      "source": [
+        "import pennylane as qml\n",
+        "from jax import numpy as jnp\n",
+        "import jax\n",
```
Comment (dime10) [inline demos/qml_optimization.ipynb:34]: I believe this import is unused

File: demos/qml_optimization.ipynb (hunk 3 above)
Comment (maliasadi) [reply demos/qml_optimization.ipynb:138]: really nice to see all these features in one demo :star_struck:

File: demos/qml_optimization.ipynb [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+        "- Compute the loss function (so we can track training) and the gradients (so we can apply an optimization step). We can do this in one execution via the `jax.value_and_grad` function.\n",
+        "\n",
+        "- Apply the update step of our optimizer via `opt.update`\n",
+        "\n",
+        "- Update the parameters via `optax.apply_updates`\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 35,
```
Comment (maliasadi) [inline demos/qml_optimization.ipynb:224]: Would you re-run this notebook to get all these execution counts in order?

File: demos/qml_optimization.ipynb (hunk 2 above)
Comment (dime10) [reply demos/qml_optimization.ipynb:85]: The problem appears to be the use of the nested `qjit` decorators.

While it is supported in the sense that inner qjit decorators are supposed to be transparent, it appears that in conjunction with certain options this causes issues (here autograph). If the support were working as intended though, the inner `autograph=True` flag would be fully ignored anyways.

Unfortunately, it appears natural for users to place `qjit` all over the place as I've seen the pattern before, whereas it really is it best thought of as the (unique!) entry point marker to the compiled program.

This could be an indication that the feature I suggested yesterday is required to better match user expectations @josh146

File: demos/qml_optimization.ipynb [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+    {
+      "cell_type": "code",
+      "execution_count": 46,
+      "metadata": {
+        "id": "I9Yaeyer9rVb"
+      },
+      "outputs": [],
+      "source": [
+        "def optimization_noqjit(params):\n",
+        "    opt = optax.adam(learning_rate=0.3)\n",
```
Comment (maliasadi) [inline demos/qml_optimization.ipynb:398]: @josh146 you may want to re-run this after removing the `optax.adam` construct from this function for a fair comparison.

**Review by maliasadi — APPROVED** [review]
Nice demo :100:

File: demos/qml_optimization.ipynb (hunk 2 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:85]: > This example seems to work fine with autograph and vmap, there must be another element 🤔

Oh yes! I think @maliasadi also noticed this yesterday on slack. It only occurs if you directly stack `qjit(vmap(func), autograph=True)`. 

If you instead do `qjit(func, autograph=True)`, where `func` internally calls `vmap`, this works fine.

Ah, I guess what is happening here is that since `vmap` is being passed to `qjit`, autograph is directly applying to `vmap` as the entry point?

File: doc/index.rst [hunk 11]
```diff
@@ -89,6 +89,7 @@ Catalyst
    demos/magic_distillation_demo
    demos/adaptive_circuits_demo
    demos/qml/tutorial_qubit_rotation
+   QML Optimization <demos/qml_optimization>
```
Comment (maliasadi) [inline doc/index.rst:92]: As this demo is using `Optax` for optimizing the workflow, would you like to add this to the index title
 
```suggestion
   QML Optimization with Optax <demos/qml_optimization>
```

File: demos/qml_optimization.ipynb (hunk 2 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:85]: In this particular case (for me as a user), since I was just trying to do `qjit(vmap(circuit))` to test if my circuit executed correctly, it seemed much more natural to me than having an additional level of function nesting 😢

File: demos/qml_optimization.ipynb (hunk 7 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:66]: oh... that was purely user error 😆 Yeah I'll be consistent here.

File: demos/qml_optimization.ipynb (hunk 5 above)
Comment (maliasadi) [reply demos/qml_optimization.ipynb:377]: make sense! as jax + default.qubit use the JAX + XLA toolchain without any callbacks while lightning.qubit uses callbacks..

File: demos/qml_optimization.ipynb (hunk 2 above)
Comment (maliasadi) [reply demos/qml_optimization.ipynb:85]: yes, that is a quick workaround for this issue by using `qjit` as the decorator instead. As being said, the root of this issue is about AG trying to convert all control flow operations to QJIT compatible ones. I also tried to use function wrappers around these if-conditions and trigger them as function calls (to see the AG behaviour in the case of nested func calls) but it didn't fix the issue.

Comment (josh146) [reply demos/qml_optimization.ipynb:85]: @maliasadi I was chatting to @dime10, and adding `vmap` to this list here might solve this issue:

https://github.com/PennyLaneAI/catalyst/blob/3606aec53842552919c19e65bbad7feb72daf278/frontend/catalyst/ag_primitives.py#L523-L531

Comment (dime10) [reply demos/qml_optimization.ipynb:85]: So unfortunately this situation is not covered by that hotfix. The big issue is the "primary" function we are feeding into qjit is an internal wrapper, and Catalyst internal functions should typically not be converted. On the other hand, the function fed into qjit must be converted by autograph since that is the starting. I think we can work around the issue with some small-ish amount of work, but for this PR we can close this comment chain I think.

Comment (dime10) [reply demos/qml_optimization.ipynb:85]: In that case removing the flag might make sense:
```suggestion
        "circuit = qml.qjit(catalyst.vmap(circuit, in_axes=(1, None)))"
```

File: demos/qml_optimization.ipynb (hunk 5 above)
Comment (dime10) [reply demos/qml_optimization.ipynb:377]: @maliasadi What kind of callbacks are you referring to?

Comment (dime10) [reply demos/qml_optimization.ipynb:377]: I think Josh is just comparing "jax.jit + default.qubit" versus "lightning.qubit" (no jax).

Comment (josh146) [reply demos/qml_optimization.ipynb:377]: yep :) Since the classical processing was quite 'light' in this case, I wasn't sure if jax + lightning.qubit was worth the comparison?

File: demos/qml_optimization.ipynb (hunk 8 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:34]:
```suggestion
```

File: demos/qml_optimization.ipynb (hunk 9 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:224]: yep can do

File: demos/qml_optimization.ipynb (hunk 6 above)
Comment (josh146) [reply demos/qml_optimization.ipynb:217]: @rmoyard I'll merge in this PR today; do you mind updating this demo in #332 before merging it in?

File: demos/qml_optimization.ipynb (hunk 5 above)
Comment (dime10) [reply demos/qml_optimization.ipynb:396]: Agreed, I don't think it would do much.

Regarding the JAX result, it is really impressive how much computation can be optimized and fused once you have access to the simulator implementation!

But 1 min for `lightning.qubit` seems shocking: ~60x over the compiled version where the optimization loop is not even compiled? :O

To answer your first question, I was just thinking `default.qubit` with no jax (for "raw PennyLane").

File: requirements.txt [hunk 12]
```diff
@@ -23,3 +23,4 @@ nbmake
 pennylane-lightning[kokkos]
 amazon-braket-pennylane-plugin>=1.23.0
 lark
+optax
```
Comment (josh146) [inline requirements.txt:26]: Hey @maliasadi @dime10, any idea why this is not fixing the CI issue with the new demo? 🤔 

https://github.com/PennyLaneAI/catalyst/actions/runs/8362374850/job/22893066971?pr=612#step:11:36

As far as I can tell, `requirements.txt` is used for this workflow.

Comment (josh146) [reply requirements.txt:26]: solved!
